**Fill the Code blocks according to the comments provided as hints**

In [2]:
from sklearn.datasets import load_diabetes #We will use this dataset

#import libraries for numpy, Linear regression, r2_score, train_test_split
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split

In [4]:
X,y = load_diabetes(return_X_y=True)  #loading dataset

In [5]:
#print shape of X and y and write in a text block what do you observe about both 
print("shape of X is", X.shape)
print("shape of y is", y.shape )


shape of X is (442, 10)
shape of y is (442,)


We observed that X has 442 samples for 10 independent features or variables and corresponding to each row - we have one dependent variable or target variable i.e. y is one dimensional with 442 elements.

In [6]:
#split the dataset into Train(80%) and Test(20%)
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.2, random_state = 2)



In [7]:
#fit X,y in Linear Regression Model of sklearn
lr = LinearRegression()
lr.fit(X_train,y_train)




LinearRegression()

In [8]:
#print coefficients and intercepts (in @d gradient descent in was only m and b but as it is n dimensional it will have n coefficients and 1 intercept)
print(lr.coef_)
print(lr.intercept_)



[  -9.15865318 -205.45432163  516.69374454  340.61999905 -895.5520019
  561.22067904  153.89310954  126.73139688  861.12700152   52.42112238]
151.88331005254167


In [9]:
#predict on X_test and calculate r2_score
y_pred = lr.predict(X_test)
r2_score(y_test, y_pred)



0.4399338661568968

In [10]:
#print X_train shape
print(X_train.shape)


(353, 10)


In [11]:
#Write GDregressor class


class GDRegressor:
    
    def __init__(self,learning_rate,epochs):
        
        self.coef_ = None
        self.intercept_ = None
        self.lr = learning_rate
        self.epochs = epochs
        
    def fit(self,X_train,y_train):
        # init your coefs
        self.intercept_ = 100         #initialise the intercept
        self.coef_ = np.ones(X_train.shape[1]) #initialise the array of coefficients according to number of coefficients
        
        for i in range(self.epochs):
            # update all the coef and the intercept
            y_hat = np.dot(X_train,self.coef_) + self.intercept_                         #apply prediction formula i.e. (y_hat = coefficient*X + intercept) in terms of matrices coefficient and X   
            intercept_derivative = -2*np.sum(y_train -y_hat)                              #find derivative matrix 
            self.intercept_ =  self.intercept_ - self.lr*(intercept_derivative)     #update intercept using above intercept_derivative
            
            coef_derivative =   -2*np.dot((y_train -y_hat),X_train)                       #derivative matrix for all coefficients in a single step
            self.coef_ =  self.coef_ -  (self.lr)*(coef_derivative)                                                    #update intercept using coef_derivative
        
        #print the coefficients array and the intercept
        print(self.coef_,self.intercept_)
        print(self.coef_.shape, X_test.shape)
    
    def predict(self,X_test):
        return np.dot (self.coef_,X_test.T)+ self.intercept_ #new prediction in the form of (y = coefficients*X + intercept) using matrices

In [12]:
#call GDRegressor by keeping epochs and learning rate according to you and keep changing it for better results
gd = GDRegressor(0.002,1000)


In [13]:
#fit X_train, y_train into GDRegressor object using function of the class
gd.fit(X_train,y_train)


[   3.01547544 -194.78060439  516.57653401  335.02703877  -50.74256462
 -127.98525807 -192.71783769   94.58594764  486.33213457   80.94601818] 151.9698323273896
(10,) (89, 10)


In [14]:
#predict on X_test using function of the class
y_pred = gd.predict(X_test)

In [15]:
#calculate r2_score to verify
r2_score(y_test,y_pred)


0.450527059505485